lib/std/debug.zig

```diff
@@ -225,23 +530,51 @@ pub fn assert(ok: bool) void {
 
 pub fn panic(comptime format: []const u8, args: anytype) noreturn {
     @setCold(true);
-    // TODO: remove conditional once wasi / LLVM defines __builtin_return_address
-    const first_trace_addr = if (builtin.os.tag == .wasi) null else @returnAddress();
-    panicExtra(null, first_trace_addr, format, args);
+
+    panicExtra(null, format, args);
+}
+
+/// `panicExtra` is useful when you want to print out an `@errorReturnTrace`
+/// and also print out some values.
+pub fn panicExtra(
+    trace: ?*builtin.StackTrace,
+    comptime format: []const u8,
+    args: anytype,
+) noreturn {
+    @setCold(true);
+
+    const size = 0x1000;
+    const trunc_msg = "(msg truncated)";
+    var buf: [size + trunc_msg.len]u8 = undefined;
+    // a minor annoyance with this is that it will result in the NoSpaceLeft
+    // error being part of the @panic stack trace (but that error should
+    // only happen rarely)
+    const msg = std.fmt.bufPrint(buf[0..size], format, args) catch |err| switch (err) {
+        std.fmt.BufPrintError.NoSpaceLeft => blk: {
+            std.mem.copy(u8, buf[size..], trunc_msg);
+            break :blk &buf;
+        },
+    };
+    builtin.panic(msg, trace);
 }
 
 /// Non-zero whenever the program triggered a panic.
-/// The counter is incremented/decremented atomically.
-var panicking: u8 = 0;
-
-// Locked to avoid interleaving panic messages from multiple threads.
+/// Counts the number of threads which are waiting on the `panic_mutex`
+/// to print out the stack traces and msg.
+var panicking: std.atomic.Int(u32) = std.atomic.Int(u32).init(0);
```

Comment:
Completely useless but otherwise correct.

```suggestion
var panicking = std.atomic.Int(u32).init(0);
```

> based on some discussions in discord

Where do I find some public logs of this lovely discussion? :)
Please summarize the rationale behind the changes.

Reply:
see https://github.com/ziglang/zig/pull/8228#discussion_r598006277